### PR TITLE
feat: restructure `ibc` crate

### LINKED ISSUE #697: Re-export all crates that have types present in our public API
For example, our public API requires `ibc_proto::google::protobuf::Any`. We should thus re-export `ibc_proto` in our `lib.rs`.

This will have many benefits:

1. When users upgrade their `ibc-rs` dependency (not through `cargo update`), it will just work, as opposed to showing a bunch of errors which effectively mean "upgrade ibc-proto as well".
2. It will make the `ClientState` derive macro (implemented in #683) slightly cleaner (i.e. users will be allowed to rename their `ibc_proto` without breaking the macro).

### LINKED ISSUE #716: Move transfer application and tendermint client to separate crates
We should move the `transfer` application and the `tendermint` client to separate crates. This way, these can evolve independently from the core ibc-rs. 

Notably, post ibc-rs 1.0, this will mean that if needed we will be able to make breaking changes to the app or client (incrementing the major version) without needing to increment ibc-rs's major version. This will completely decouple core IBC from Tendermint.

### LINKED ISSUE #771: rename `{submodule}.rs` with corresponding `{submodule}` directory to `{submodule}/mod.rs`
## Improvement Summary

Rename all `{submodule}.rs` with a corresponding `{submodule}/` to `{submodule}/mod.rs`. This doesn't change anything in the rust namespace.

Example.

`crates/ibc/src/applications/transfer/msgs/` and `crates/ibc/src/applications/transfer/msgs.rs` exist. So, I propose to rename the rust file to `crates/ibc/src/applications/transfer/msgs/mod.rs`.

## Proposal

Usually the file explorers sort the directories first and then the files. So, maintaining `{submodule}.rs` and `{submodule}/` directory makes them separated on the viewing list.

`fd -e rs -x ls -d {.} 2> /dev/null` provides a list of directories which falls under this.

### LINKED ISSUE #948: Remove redundant #[test_log::test] attributes in test modules
### Description
There are places in our test modules (e.g. [here](https://github.com/cosmos/ibc-rs/blob/1ec4b8ca4df85175acb75189c8d47bc8d0ba7178/crates/ibc/src/core/ics24_host/identifier/validate.rs#L147)), we have used `#[test_log::test]` as the test attribute. However, neither the `ibc-rs` implementation nor its tests involve any logging or tracing. We should assess the codebase to identify and eliminate any instances where these attributes are redundant.

### LINKED ISSUE #965: Restructure the `ibc` crate
## Background
This initiative aims to optimize the organization of the `ibc` crate by decomposing it into smaller, self-contained libraries. A pivotal step in this process involved segregating our mock testing facilities into the standalone `ibc-testkit` library, as detailed in [this PR](https://github.com/cosmos/ibc-rs/pull/954). By decoupling its implementation from the main `ibc` crate, we've laid the groundwork for advancing this initiative.

The primary objectives here are twofold: firstly, to diminish the interdependence within the codebase among various components such as ibc core, clients, and applications, and secondly, to elevate the overall usability of the implementation. The overarching aim is to empower users to selectively import specific IBC layers, mitigating potential conflicts related to dependencies or features that may arise in the context of a monolithic library.

## Steps
The following steps outline the restructuring process:

```[tasklist]
### Tasks
- [x] Write an ADR explaining why this change is necessary and how the codebase will look like afterward
- [x] Restructure and split off `applications` module under `ibc-apps`
- [x] Restructure and split off light clients under `ibc-clients`
- [x] Restructure and split off `core` module under the `ibc-core`
- [x] Define a new library for base types and preludes imported by other crates
- [x] Move main directories/libraries to the root of the repository, eliminating the need for the `/crates` directory.
- [ ] https://github.com/cosmos/ibc-rs/issues/976
```

## Next
Following the above steps, the next phase involves preparing a GitHub action for streamlining the release process:
- https://github.com/cosmos/ibc-rs/issues/983

## Checklist
Additionally, for each added library, we should adhere to the following checklist:

- Use workspace dependency
- Remove unused dependencies and feature flags
- Add a README and short description for each library
- Update the main README page
- Update the Releases section of `contributing.md`
- Reflect changes on basecoin-rs

## Relevant issues
- [x] https://github.com/cosmos/ibc-rs/issues/716
- [x] https://github.com/cosmos/ibc-rs/issues/552

### LINKED ISSUE #976: Expose different consts, types and impls for public access
## Background

Part of #965

To enable the `ibc` restructure and also maximize users' access to our implementation, we'll be making the following types and implementations public:

- `validate` and `execute` handler functions for all the IBC message types.
-  `TYPE_URL` constants.
- Any private fields within the domain message types.
- Any private fields within the Tendermint `ClientState` and `ConsensusState` structs.

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #976
Closes: #965
Closes: #948
Closes: #771
Closes: #716
Closes: #697

Integration tests: https://github.com/informalsystems/basecoin-rs/pull/144

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-core/README.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# IBC Core
+
+This crate is top-level library re-exports implemented Inter-Blockchain
+Communication (IBC) core modules serves as a meta-crate, simplifying the
+process of importing and integrating various IBC core modules into your
+blockchain. IBC is a distributed protocol that enables communication between
+distinct sovereign blockchains and IBC core is the part of the protocol that
+handles transport, authentication and ordering (TAO) layers of data packets.
```
Comment (seanchen1991) [inline ibc-core/README.md]:
```suggestion
This is the top-level library that re-exports the Inter-Blockchain Communication (IBC) core modules as a meta-crate. It serves to simplify the process of importing and integrating various IBC core modules into your blockchain.

IBC is a distributed protocol that enables communication between distinct sovereign blockchains and IBC core is the part of the protocol that handles the transport, authentication, and ordering (TAO) of data packets.
```

File: ibc-core/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+The structure within the `ibc-core` crate is designed to provide flexibility for
+external users. It allows you to utilize the own `ibc-core` crate or selectively
+import specific libraries, whether you need a certain IBC core module (e.g.
+`ibc-core-client` crate) or only their associated data structures (e.g.
+`ibc-core-client-types`). This versatility empowers hosts, including chain
+integrators, relayers, or any IBC tooling projects, to build their solutions on
+top of the layers that best suit their requirements.
```
Comment (seanchen1991) [inline ibc-core/README.md]:
```suggestion
external users. You can choose to utilize the entire `ibc-core` crate, or selectively import specific libraries. From there, you also have the flexibility of bringing in an entire sub-module (e.g. the `ibc-core-client` crate), or only a module's associated data structures (e.g. `ibc-core-client-types`).

This versatility empowers hosts, including chain integrators, relayers, or any IBC tooling projects, to build their solution on top of the layers that best suit their particular requirements.
```

File: ibc-derive/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# IBC Derive
+
+This crate provides a set of procedural macros for deriving IBC-related traits
+defined in the [`ibc-rs`](https://github.com/cosmos/ibc-rs) repository. reducing
+the amount of boilerplate code required to implement them. It already contains
+macro implementations for the following traits:
```
Comment (seanchen1991) [inline ibc-derive/README.md]:
```suggestion
This crate provides a set of procedural macros for deriving IBC-related traits
defined in the [`ibc-rs`](https://github.com/cosmos/ibc-rs) repository, reducing
the amount of boilerplate code required to implement them. It contains
macro implementations for the following traits:
```

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -5,15 +5,31 @@ on:
...
-      - crates/**
+      - ibc/**
+      - ibc-core/**
+      - ibc-apps/**
+      - ibc-data-types/**
+      - ibc-clients/**
+      - ibc-primitives/**
+      - ibc-query/**
+      - ibc-testkit/**
+      - ibc-derive/**
```
Comment (rnbguy) [inline .github/workflows/rust.yml:16]: I think you can do `ibc*/**`, unless you want to be selective.

File: ibc-apps/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# IBC Applications
+
+This crate is a top-level library that re-exports Inter-Blockchain Communication
+(IBC) applications implemented. It serves as a meta-crate, simplifying the
+process of importing and integrating various IBC applications into your
+blockchain. IBC is a distributed protocol that enables communication between
+distinct sovereign blockchains. IBC applications abstract away the core
+transport, authentication and ordering (TAO) layers, letter blockchain app
+developers focus solely on business logic implementation.
```
Comment (seanchen1991) [inline ibc-apps/README.md]:
```suggestion
This crate is a top-level library that re-exports implementations of Inter-Blockchain Communication
(IBC) applications. It serves as a meta-crate, simplifying the
process of importing and integrating various IBC applications into your
blockchain. IBC is a distributed protocol that enables communication between
distinct sovereign blockchains. IBC applications abstract away the core
transport, authentication, and ordering (TAO) layers, letting blockchain app
developers focus solely on implementing business logic.
```

File: .github/workflows/rust.yml (hunk 4 above)
Comment (Farhad-Shabani) [reply .github/workflows/rust.yml:16]: yea, still being able to selectively comment out each crates works better for the development and debugging purposes imo.

File: ibc-clients/README.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# IBC Clients
+
+This crate is top-level library re-exports implemented Inter-Blockchain
+Communication (IBC) light clients serves as a meta-crate, simplifying the
+process of importing and integrating various IBC clients into your blockchain.
+IBC is a distributed protocol that enables communication between distinct
+sovereign blockchains and IBC light clients track the consensus states and proof
+specs of external blockchains required to properly verify proofs against the
+client's consensus state.
```
Comment (seanchen1991) [inline ibc-clients/README.md]:
```suggestion
This top-level crate re-exports Inter-Blockchain
Communication (IBC) implementations of light clients. It serves as a meta-crate, simplifying the
process of importing and integrating various IBC clients into your blockchain.
IBC is a distributed protocol that enables communication between distinct
sovereign blockchains and IBC light clients track the consensus states and proof
specs of external blockchains, which are required to properly verify proofs against the
client's consensus state.
```

File: ibc-clients/ics07-tendermint/src/client_state.rs [hunk 7]
```diff
@@ -0,0 +1,574 @@
+use ibc_client_tendermint_types::error::Error;
```
Comment (seanchen1991) [inline ibc-clients/ics07-tendermint/src/client_state.rs:11]:
```suggestion
//! This module includes trait implementations for the `ibc_client_tendermint_types::ClientState` 
//! type. Implemented traits include `ClientStateCommon`, 
//! `ClientStateValidation`, and `ClientStateExecution`. 
//! 
//! Note that this crate defines a newtype wrapper around the `ibc_client_tendermint_types::ClientState` 
//! type in order to enable implementing a foreign trait on a 
//! foreign type (i.e. the orphan rule in Rust). As such, this 
//! module also includes some trait implementations that serve 
//! to pass through traits implemented on the wrapped `ClientState` 
//! type.

use ibc_client_tendermint_types::error::Error;
```

File: ibc-clients/ics07-tendermint/src/consensus_state.rs [hunk 8]
```diff
@@ -0,0 +1,90 @@
+use ibc_client_tendermint_types::error::Error;
```
Comment (seanchen1991) [inline ibc-clients/ics07-tendermint/src/consensus_state.rs:9]:
```suggestion
//! This module includes trait implementations for the `ibc_client_tendermint_types::ConsensusState` 
//! type. It implements the `ConsensusStateTrait` for 
//! `ConsensusState` by defining a newtype wrapper in order
//! to circumvent Rust's orphan rule, which disallows foreign
//! traits from being implemented on foreign types.
//! This module also includes some trait implementations that serve 
//! to pass through traits implemented on the wrapped
//! `ConsensusState` type.

use ibc_client_tendermint_types::error::Error;
```

**Review by seanchen1991 — APPROVED** [review]

**Pull request merged**